src/Algebra/Graph/Algorithm/Internal.hs

```diff
@@ -1,18 +1,77 @@
+{-# LANGUAGE FlexibleContexts #-}
 module Algebra.Graph.Algorithm.Internal where
 
-import                            Algebra.Graph
-import Data.DisjointSet           (DisjointSet)
-import qualified Data.DisjointSet as DisjointSet
+import           Algebra.Graph
+import           Control.Monad
+import           Control.Monad.ST
+import           Data.Map.Strict   (Map)
+import qualified Data.Map.Strict   as Map
+import           Data.UnionFind.ST (Point)
+import qualified Data.UnionFind.ST as UF
 
--- | Extract disjoint set of connectivity components from the graph
+-- | O(s + n log n)
+-- Extract disjoint set of connectivity components from the graph
 --
--- >>> DisjointSet.pretty $ components ((1 * 2) + (3 * 4))
--- "{{1,2},{3,4}}"
+-- **NOTE:** Doesn't work properly. Check an example.
 --
--- >>> DisjointSet.pretty $ components (((1 * 2) + (3 * 4)) + (2 * 3))
--- "{{1,2,3,4}}"
-components :: Ord a => Graph a -> DisjointSet a
-components Empty           = DisjointSet.empty
-components (Vertex x)      = DisjointSet.singleton x
-components g@(Connect _ _) = DisjointSet.singletons $ vertexSet g
-components (Overlay x y)   = components x <> components y
+-- >>> components (((1 * 2) + (3 * 4)) + (2 * 3))
+-- fromList [(2,[2,1]),(4,[4,3])]
+components :: Ord a => Graph a -> Map a [a]
+components g = runST $ do
+  (ps, g') <- mkVertexPoints g Map.empty
+  componentsST g'
+  componentsFromPoints $ Map.toList ps
+
+-- | O(n log n). Not sure.
+componentsFromPoints :: Ord a => [(a, Point s a)] -> ST s (Map a [a])
+componentsFromPoints vs = do
+  list <- pointsList vs
+  return $ Map.fromListWith (<>) list
+
+-- | O(n).
+pointsList :: [(a, Point s a)] -> ST s [(a, [a])]
+pointsList = mapM $ \(x, px) -> do
+  repr <- UF.descriptor px
+  return (repr, [x])
+
+-- | O(s + n log n).
+mkVertexPoints :: Ord a => Graph a -> Map a (Point s a) -> ST s (Map a (Point s a), Graph (Point s a))
+mkVertexPoints Empty _ = return (Map.empty, Empty)
+mkVertexPoints (Vertex v) ps =
+  case Map.lookup v ps of
+    Nothing -> do
+      p <- UF.fresh v
+      return (Map.insert v p ps, Vertex p)
+    _ -> return (ps, Empty)
+mkVertexPoints (Connect l r) ps = do
+  (lp, lg) <- mkVertexPoints l ps
+  (rp, rg) <- mkVertexPoints r ps
+  return (lp <> rp, Connect lg rg)
+mkVertexPoints (Overlay l r) ps = do
+  (lp, lg) <- mkVertexPoints l ps
+  (rp, rg) <- mkVertexPoints r ps
+  return (lp <> rp, Overlay lg rg)
```

Comment:
State should be threaded. When you consider right subgraph, you should account for points that have been created on the left:

```suggestion
  (lp, lg) <- mkVertexPoints l ps
  (rp, rg) <- mkVertexPoints r lp
  return (rp, Overlay lg rg)
```